keps/sig-storage/177-volume-snapshot/README.md

```diff
@@ -231,24 +251,329 @@ Here are the changes since the original design proposal:
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+  Describe manual testing that was done and the outcomes.
+  Longer term, we may want to require automated upgrade/rollback tests, but we
+  are missing a bunch of machinery and tooling and can't do that now.
+  We plan to do this testing after the code is implemented that moves snapshot API from v1beta1 to v1. We will test the following:
+  * Upgrade from v1beta1 to v1. Take a snapshot and create a PVC from a snapshot. Also test delete snapshot and PVC.
+  * Downgrade from v1 to v1beta1. Check the snapshot and PVC created above. Take a snapshot and create a PVC from a snapshot. Also test delete snapshot and PVC.
+  * Upgrade from v1beta1 to v1 again. Check the snapshot and PVC created earlier. Take a snapshot and create a PVC from a snapshot. Also test delete snapshot and PVC.
+  * Also verify snapshots taken with v1beta, can be:
+    * restored using v1 + v1beta1
+    * deleted using v1 + v1beta1
+
+* **Is the rollout accompanied by any deprecations and/or removals of features, APIs,
+fields of API types, flags, etc.?**
+  Even if applying deprecation policies, they may still surprise some users.
+
+  We have added a validation webhook that will tighten the API validation. See details in "keps/sig-storage/1900-volume-snapshot-validation-webhook".
+  This validating admission webhook should be applied first before going GA to prevent invalid API objects from being created. If the webhook is not applied before going GA, it means there could be invalid API objects created. It also means existing v1beta1 invalid objects will error on any update, including delete. A distribution/admin must not add v1 API until they are sure no invalid v1beta1 objects exist.
+
+  The feature gate is only for creating PVC from a volume snapshot data source. The snapshot controller and validating webhook service are out-of-tree controllers which implement the snapshot feature, and their lifecycle/management/rollback is irrelevant to feature gate. The validation webhook is validating VolumeSnapshot and VolumeSnapshotContent API objects. So if the validation webhook is installed and then the feature gate is disabled, it will just prevent PVC from being created from a snapshot.
+
+### Monitoring Requirements
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can an operator determine if the feature is in use by workloads?**
+  Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+  checking if there are objects with field X set) may be a last resort. Avoid
+  logs or events for this purpose.
+  Metrics support is already added to CSI external-snapshotter sidecar. Metrics support will be added to the snapshot-controller in 1.20.
+  We have the following metrics:
+  snapshot_operation_total_seconds, snapshot_operation_count (indicating success/failure), snapshot_csi_total_seconds
```

Comment:
@xing-yang 
per discussion, please remove this metrics, and add: csi_sidecar_operations_seconds
original idea was following - 
snapshot_operation_total_seconds: end to end latency metric for deletion/creation of a VolumeSnapshot CR, reported from snapshot controller
snapshot_csi_total_seconds: end to end latency metric for deletion/creation/listing CSI calls from sidecar. 
csi_sidecar_operations_seconds: latency metric for ONE CSI call

One of the things I was thinking is to be able to tell how much time has been spent on CSI calls waiting v.s. how much time controllers takes to populate status etc. snapshot_operation_total_seconds - snapshot_csi_total_seconds should answer that question. However, the differences between these two should be insignificant. @msau42

Reply:
Sure, I'll remove `snapshot_csi_total_seconds`.